src/tagsQLController.js

```diff
@@ -0,0 +1,395 @@
+        } else if (needOneRightOperand && currentRightOperands > 1) {
+          // Remove excedent
+          segments.splice(index+2, currentRightOperands-1);
+        } else if (!needOneRightOperand && !isEnum && currentRightOperands > 0) {
+          // Remove values
+          segments.splice(index+1, currentRightOperands);
+        } else if (isEnum) {
+          if (currentRightOperands === 0) {
+            // Add a value and plus-button
+            segments.splice(index+1, 0, this.uiSegmentSrv.newKeyValue('select value'), this.uiSegmentSrv.newPlusButton());
+          } else if (currentRightOperands === 1) {
+            // Just add plus-button after the value
+            segments.splice(index+2, 0, this.uiSegmentSrv.newPlusButton());
+          }
+        }
+      }
+      segments[index] = segment;
+    }
+    this.targetSupplier().tagsQL = segmentsToString(segments);
+  }
+}
+
+export function convertFromKVPairs(kvTags) {
+  if (kvTags === undefined) {
+    return undefined;
+  }
+  return kvTags.map(tag => {
+    if (tag.value === '*' || tag.value === ' *') {
+      return tag.name;
+    }
+    if (tag.value.charAt(0) === '$') {
+      // it's a variable
+      return tag.name + " IN [" + tag.value + "]";
+    }
+    return tag.name + "='" + tag.value + "'";
+  }).join(' AND ');
+}
+
+function valueToString(value) {
+  if ((value.charAt(0) === "'" && value.charAt(value.length-1) === "'")
+      || value.match(/^\$?[a-zA-Z0-9]+$/g)) {
+    // Variable, simple literal or already single-quoted => keep as is
+    return value;
+  }
+  return "'" + value + "'";
+}
+
+export function segmentsToString(segments) {
+  var strTags = "";
+  var i = 0;
+  while (i < segments.length) {
+    if (segments[i].type === 'plus-button') {
+      i++;
+      continue;
+    }
+    if (i != 0) {
+      // AND/OR
+      strTags += " " + segments[i++].value + " ";
+    }
+    // Tag name
+    let tagName = segments[i++].value;
+    // Operator
+    let op = segments[i++].value;
+    if (op === OPERATOR_EQ || op === OPERATOR_NOTEQ) {
+      strTags += tagName + op + valueToString(segments[i++].value);
+    } else if (op === OPERATOR_EXISTS) {
+      strTags += tagName;
+    } else if (op === OPERATOR_NOTEXISTS) {
+      strTags += 'NOT ' + tagName;
+    } else if (op === OPERATOR_IN) {
+      let v = valuesToString(segments, i);
+      i = v.i;
+      strTags += tagName + ' IN [' + v.values + ']';
+    } else if (op === OPERATOR_NOTIN) {
+      let v = valuesToString(segments, i);
+      i = v.i;
+      strTags += tagName + ' NOT IN [' + v.values + ']';
+    }
+  }
+  return strTags;
+}
+
+function valuesToString(segments, i) {
+  var values = "";
+  var sep = "";
+  while (i < segments.length && segments[i].type === 'value') {
+    values += sep + valueToString(segments[i++].value);
+    sep = ",";
+  }
+  return {
+    values: values,
+    i: i
+  };
+}
+
+export function stringToSegments(strTags, segmentFactory) {
```

Comment:
Same as above, feels overcomplicated, care to share JSON example so we can think of something more readable?